### DISCUSSION TITLE
I would like to know about the behavior of the server's Parallel option. Doesn't processing in parallel = increase GPU usage?

### DISCUSSION THREAD

**Discussion by Taikono-Himazin** [ggml-org/llama.cpp#6782] [Q&A]
I would like to know about the behavior of the server's Parallel option. Doesn't processing in parallel = increase GPU usage?

Suppose that the behavior without the parallel option is as follows.
![image](https://github.com/ggerganov/llama.cpp/assets/25480502/7fcfcc6e-1601-4ef2-a3ea-755548c87526)

I thought that adding the parallel option would work as follows.
![image](https://github.com/ggerganov/llama.cpp/assets/25480502/e31ca812-d096-48be-8ca5-7cc328459f90)

But in reality it seems to work like this. 
![image](https://github.com/ggerganov/llama.cpp/assets/25480502/40f50f4d-0761-4df4-a509-0704f44b131e)

Even with the parallel option, GPU usage does not increase. Throughput has also increased only slightly. (On my server, GPU usage stays at around 10%)

Is my understanding correct?

**Comment by Taikono-Himazin** [comment]
Is it possible that parallel only works in streaming mode?

**Reply by phymbert** [reply]
No it does not increase the VRAM as each sequence uses a part of the unified KV cache. i.e n_slot_ctx = n_ctx / parallel.

It works well also with streaming enabled.

Have you faced any issue ?

**Reply by Taikono-Himazin** [reply]
I would like to increase throughput.
Current GPU usage is around 10%.
So, I tried to perform 10 parallel processes and get 10 times the current throughput. For this purpose, I have set the following options:
`--Parallel 10 -cb`
But that hasn't actually happened. Why?

**Reply by phymbert** [reply]
Please share how the steps you have followed, from HuggingFace to the server config and http requests you are sending.

You can use the following discussion as a reference, on how to configure the server for performances
- #6730

**Reply by Taikono-Himazin** [reply]
Thank you for the information! That helps a lot!
I'll check it now.
Does this mean that the llama.cpp server is not suitable if performance is required?

**Reply by phymbert** [reply]
not at all, performances are really good on small devices, or on metal, cpu backends. But the CUDA backend is currently in performance improvment devlopment, see for example:

- #5021 
- #6763 

The server is kind of production ready:

- #6398
